File: built-in plugins dsp/mb-synth-supersaw.cpp
/**
 * MB Supersaw
 * Category : instrument
 * Type     : analog
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Massive supersaw stack
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SYNTH_SUPERSAW_H
#define MB_SYNTH_SUPERSAW_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSynthSupersaw : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-synth-supersaw";
    static constexpr const char* PLUGIN_NAME    = "MB Supersaw";
    static constexpr const char* PLUGIN_TYPE    = "analog";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float detune = 15f;  // range [0, 50]
    float volume = 0.75f;  // range [0, 1]
    };

    MbSynthSupersaw() = default;
    ~MbSynthSupersaw() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.detune = std::clamp(params.detune, 0f, 50f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Supersaw
        return input;
    }
};

#endif // MB_SYNTH_SUPERSAW_H

File: built-in plugins dsp/mb-synth-supersaw.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbSynthSupersawPlugin: PluginDefinition = {
  id: "mb-synth-supersaw",
  slug: "mb-synth-supersaw",
  name: "MB Supersaw",
  category: "instrument",
  type: "analog",
  version: "1.0.0",
  description: "Massive supersaw stack",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: -15, gain: 0.15 },
    { type: "sawtooth", detune: -7, gain: 0.15 },
    { type: "sawtooth", detune: 0, gain: 0.2 },
    { type: "sawtooth", detune: 7, gain: 0.15 },
    { type: "sawtooth", detune: 15, gain: 0.15 },
  ],
  envelope: { attack: 0.02, decay: 0.4, sustain: 0.7, release: 0.5 },
  parameters: [
    {
      id: "detune",
      name: "Detune",
      type: "float",
      defaultValue: 15,
      minValue: 0,
      maxValue: 50,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.75,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { detune: 15, volume: 0.75 },
};

export default MbSynthSupersawPlugin;
